spanner/mutation_test.go

```diff
@@ -430,6 +432,139 @@ func TestStructToMutationParams(t *testing.T) {
+	}
+	if !testEqual(gotVals, wantVals) {
+		t.Errorf("got vals %v, want %v", gotVals, wantVals)
+	}
+}
+
+func TestReadWrite_Generated(t *testing.T) {
+	t.Parallel()
+	server, client, teardown := setupMockedTestServer(t)
+	defer teardown()
+	// The full name is generated by the server.
+	server.TestSpanner.PutStatementResult(
+		"SELECT Id, FirstName, LastName, FullName FROM Users WHERE Id = 1",
+		&StatementResult{
+			Type: StatementResultResultSet,
+			ResultSet: &sppb.ResultSet{
+				Metadata: &sppb.ResultSetMetadata{
+					RowType: &sppb.StructType{
+						Fields: []*sppb.StructType_Field{
+							{Name: "Id", Type: &sppb.Type{Code: sppb.TypeCode_INT64}},
+							{Name: "FirstName", Type: &sppb.Type{Code: sppb.TypeCode_STRING}},
+							{Name: "LastName", Type: &sppb.Type{Code: sppb.TypeCode_STRING}},
+							{Name: "FullName", Type: &sppb.Type{Code: sppb.TypeCode_STRING}},
+						},
+					},
+				},
+				Rows: []*proto3.ListValue{
+					{
+						Values: []*proto3.Value{
+							intProto(1),
+							stringProto("First"),
+							stringProto("Last"),
+							stringProto("First Last"),
+						},
+					},
+				},
+			},
+		},
+	)
+
+	type User struct {
+		ID        int64 `spanner:"Id"`
+		FirstName string
+		LastName  string
+		FullName  string `spanner:",generated"`
+	}
+	user := &User{
+		ID:        1,
+		FirstName: "First",
+		LastName:  "Last",
+	}
+	// this should work because FullName is a generated column.
+	m, err := InsertStruct("Users", user)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = client.Apply(context.Background(), []*Mutation{m})
+	if err != nil {
+		t.Fatal(err)
+	}
+	// Verify that the generated column 'FullName' was excluded from the write mutation.
+	reqs := drainRequestsFromServer(server.TestSpanner)
+	var commitReq *sppb.CommitRequest
+	for _, r := range reqs {
+		if c, ok := r.(*sppb.CommitRequest); ok {
+			commitReq = c
+		}
+	}
+	if commitReq == nil {
+		t.Fatalf("no CommitRequest captured; got %v", reqs)
+	}
+	// Find the write mutation for Users.
+	var write *sppb.Mutation_Write
+	for _, mut := range commitReq.Mutations {
+		if ins := mut.GetInsert(); ins != nil && ins.Table == "Users" {
+			write = ins
+			break
+		}
+	}
+	if write == nil {
+		t.Fatalf("no write mutation for table Users in CommitRequest: %v", commitReq.Mutations)
+	}
+	// Ensure FullName is not present in columns.
+	for _, col := range write.Columns {
+		if col == "FullName" {
+			t.Fatalf("generated column FullName must be excluded from write.Columns: %v", write.Columns)
+		}
+	}
+	wantCols := []string{"Id", "FirstName", "LastName"}
+	if !reflect.DeepEqual(write.Columns, wantCols) {
+		t.Fatalf("write.Columns mismatch\ngot  %v\nwant %v", write.Columns, wantCols)
+	}
+	if len(write.Values) != 1 {
+		t.Fatalf("write.Values length mismatch: got %d, want 1", len(write.Values))
+	}
+	if got := len(write.Values[0].Values); got != len(wantCols) {
+		t.Fatalf("write.Values[0] length mismatch: got %d, want %d", got, len(wantCols))
+	}
```

Comment:
nit: In general, these types of checks can best be written like this:

```
if g, w := len(write.Values[0].Values), len(wantCols); g != w {
  t.Fatalf("write.Values[0] length mismatch\n Got: %v\nWant: %v", g, w)
}
```

This has a couple of advantages:
1. It is less likely that you make a typo in either the if or the t.Fatalf call (e.g. in this case `len(wantCols)` is repeated in both, but a small error in a copy-paste action could easily mess that up)
2. By using a newline and aligning the ` Got` and `Want`, it is easier to spot the actual difference between the two values, especially if they are a bit long strings, and the only difference is for example a whitespace character.